db/compaction/compaction_iterator.cc

```diff
@@ -225,12 +225,28 @@ void CompactionIterator::InvokeFilterIfNeeded(bool* need_skip,
       value_.clear();
       iter_stats_.num_record_drop_user++;
     } else if (filter == CompactionFilter::Decision::kChangeValue) {
+      if (ikey_.type == kTypeBlobIndex) {
+        // value transfer from blob file to inlined data
+        ikey_.type = kTypeValue;
+        current_key_.UpdateInternalKey(ikey_.sequence, ikey_.type);
+      }
       value_ = compaction_filter_value_;
     } else if (filter == CompactionFilter::Decision::kRemoveAndSkipUntil) {
       *need_skip = true;
       compaction_filter_skip_until_.ConvertFromUserKey(kMaxSequenceNumber,
                                                        kValueTypeForSeek);
       *skip_until = compaction_filter_skip_until_.Encode();
+    } else if (filter == CompactionFilter::Decision::kChangeBlobIndex) {
+      if (ikey_.type == kTypeValue) {
+        // value transfer from inlined data to blob file
+        ikey_.type = kTypeBlobIndex;
+        current_key_.UpdateInternalKey(ikey_.sequence, ikey_.type);
+      }
+      value_ = compaction_filter_value_;
+    } else if (filter == CompactionFilter::Decision::kIOError) {
+      status_ =
+          Status::IOError("Failed to access blob during compaction filter");
+      valid_ = false;
```

Comment:
Yes, only valid_ == true would break the loop. I will update and add UT to cover.

Reply:
Added unit tests, and found some defects in original implementation.
1. blob_compactoin_filter.cc, line:298 blob_file_ is reset before error log, which uses blob_file_->PathName().
2. blob_db_impl.cc, line:830, if CheckOrCreateWriterLocked fails, we need to reset blob_file, otherwise ~BlobIndexCompactionFilterBase() will call CloseAndRegisterNewBlobFile() and it hits some assert eventually (e.g. log_writer_->AppendFooter(footer)).

Reply:
Great catch, thanks @burtonli !